modules/order/src/Form/MultipleProfileTypesConfirmForm.php

```diff
@@ -0,0 +1,404 @@
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getQuestion() {
+    return $this->t('Are you sure you want to switch to using multiple
+      profile types for shipping and billing for the %label order type?', [
+        '%label' => $this->entity->label(),
+      ]
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getConfirmText() {
+    return $this->t('Switch to Multiple Profile Types');
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getCancelUrl() {
+    return $this->entity->toUrl('collection');
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildForm(array $form, FormStateInterface $form_state) {
+    // Check if we have incompatible module versions as modules like Commerce
+    // Shipping and Commerce POS will be affected when we switch to split
+    // profiles.
+    $incompatible_modules = $this->getIncompatibleModules();
+
+    // If we have modules that are running incompatible versions, output a
+    // warning message to the user.
+    if ($incompatible_modules) {
+      $this->messenger()->addWarning($this->t('The following modules are
+        running versions that are incompatible with using multiple profile types
+        and it could possibly render the site as unusable.
+        <p><strong>@modules</strong></p>
+        <p>Please upgrade and try again.</p>', [
+          '@modules' => implode('<br>', $incompatible_modules)
+        ]
+      ));
+
+      return;
+    }
+
+    $this->messenger()->addWarning($this->t('
+      If you choose to proceed, profiles for existing orders will be migrated to
+      use separate profile types. That can take some time and it might cause 
+      errors if the users try to view their existing orders during the process;
+      if you are running this on a production site please make sure that you are
+      in maintenance mode.'
+    ));
+
+    return parent::buildForm($form, $form_state);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitForm(array &$form, FormStateInterface $form_state) {
+    /** @var \Drupal\commerce_order\Entity\OrderTypeInterface $order_type */
+    $order_type = $this->entity;
+
+    // Create the billing and shipping profile types.
+    $this->createProfileTypes();
+
+    // Migrate any fields added to the customer profile type to the billing and
+    // shipping profile types.
+    $this->migrateProfileFields();
+
+    // Batch process to migrate the existing order profiles from the customer
+    // profile type to use the billing/shipping profile types.
+    $this->migrateExistingProfiles();
+
+    // Set the useMultipleProfileTypes field to TRUE now that we've processed
+    // everything.
+    $order_type->setUseMultipleProfileTypes(TRUE);
+    $order_type->save();
+
+    $form_state->setRedirectUrl($order_type->toUrl('collection'));
+  }
+
+  /**
+   * Create the billing and shipping profile types, if not already created.
+   */
+  protected function createProfileTypes() {
```

Comment:
@krystalcode Now that I'm writing the tests, I'm thinking couldn't we actually create the new billing and profile types this way:

```php
  /**
   * Create the new billing and shipping profile types.
   */
  protected function createProfileTypes() {
    $profile_types = [
      OrderType::PROFILE_BILLING => t('Customer Billing'),
      OrderType::PROFILE_SHIPPING => t('Customer Shipping'),
    ];

    // Load the 'customer' profile type so we can just duplicate that for the
    // new billing and shipping profile types.
    $profile_type_storage = \Drupal::entityTypeManager()->getStorage('profile_type');
    /** @var \Drupal\commerce_order\Entity\OrderType $order_type */
    $profile_type = $profile_type_storage->load(OrderType::PROFILE_COMMON);

    // Fetch the non-base fields from the 'customer' profile type so we can copy
    // the same to the new profile types.
    $extra_fields_to_add = [];
    $field_definitions = \Drupal::service('entity_field.manager')->getFieldDefinitions('profile', OrderType::PROFILE_COMMON);
    foreach ($field_definitions as $field_name => $field_definition) {
      if ($field_definition->getFieldStorageDefinition()->isBaseField()) {
        continue;
      }

      $extra_fields_to_add[$field_name] = $field_definition;
    }

    // Add the billing and shipping profile types duplicated from the 'customer'
    // profile type.
    foreach ($profile_types as $bundle => $label) {
      // Create the new profile type.
      $new_profile_type = $profile_type->createDuplicate();
      $new_profile_type->set('id', $bundle);
      $new_profile_type->set('bundle', $bundle);
      $new_profile_type->set('label', $label);
      $new_profile_type->save();

      // Now add the non-base fields to this new profile type.
      foreach ($extra_fields_to_add as $field_name => $field_definition) {
        $new_field_definition = $field_definition->createDuplicate();
        $new_field_definition->set('entity_type', 'profile');
        $new_field_definition->set('bundle', $bundle);
        $new_field_definition->save();
      }
    }
  }
```

Basically, just duplicate the existing 'customer' profile type. This way we don't need to add or import the config files that create the new profile types.

Reply:
@shabananavas It is possible, however there is one caveat. That will also copy any properties such as the flag that defines to use revisions etc. Can you review what else will be copied and if that is desirable?

Reply:
@krystalcode It copies the following fields over:

![image](https://user-images.githubusercontent.com/11439056/44567898-0f62d800-a792-11e8-8a77-6ddf9af3c0d7.png)

![image](https://user-images.githubusercontent.com/11439056/44568001-77192300-a792-11e8-8f98-aa5f880c9f17.png)

The ^^ last 3 are just fields I added to confirm they get created as well. It does copy the 'use_revisions' flag.

![image](https://user-images.githubusercontent.com/11439056/44567978-5e107200-a792-11e8-8e8a-1f67a5a2c01f.png)

But isn't this what we want anyway? We're setting use_revision to TRUE for the 'customer' profile type and then hiding that field from the user so that they don't change it because we don't want the order references to be deleted.